src/condor_tests/ornithology-docs/source/api/custom_plugins.rst

```diff
@@ -0,0 +1,213 @@
+These plugins are readily available for jobs to use as part of the
+default file transfer plugin list, but can also be referred to by
+:doc:`predefined`.
+
+NULL Plugin
+-----------
+
+The NULL plugin copies all data from a file located on disk to
+``/dev/null`` (or equivalent). This allows transferring of a file
+without actually copying the file elsewhere on disk.
+
+The NULL plugin accepts transfers using the ``null://`` URL schema.
+
+.. warning::
+
+    Transferred files need to exist on disk at transfer time.
+
+Debug Plugin
+------------
+
+The Debug plugin that does not transfer any files but rather decodes
+information encoded in the URL to decide how to act or what result
+ClassAds to output.
+
+The Debug plugin accepts transfers using the following URL schemas:
+
+1. ``debug://``
+2. ``decode://``
+3. ``encoded://``
+
+URL Encoding
+~~~~~~~~~~~~
+
+The URL has two levels of encoding. The first level of encoding is
+the main command immediately proceeding the URL schema followed by
+additional details which are forward slash (``/``) separated e.g.
+``debug://command/(details)``. The details may be optional depending
+on the command. The following are the available main commands:
+
+*EXIT*
+    - **Format**: ``schema://exit/<value>``
+    - **Example**: ``debug://exit/1``
+    - **Description**: Informs the plugin to exit immediately with an
+      intended exit code specified in the extra details (immediately
+      following the command). The plugin will always exit when this
+      command is specified but may not exit with the specified number
+      provided the following reasons:
+
+      1. Any previously parse URL had an invalid encoding.
+      2. The provided value was negative (exit with the absolute value).
+      3. The provided value was out of the range of ``0`` - ``123``.
+         Will choose the closest viable exit code (likely ``123``).
+
+*SLEEP*
+    - **Format**: ``schema://sleep/<value>``
+    - **Example**: ``debug://sleep/20``
+    - **Description**: Informs the plugin to sleep for the specified time
+      immediately following the command.
+
+*SIGNAL*
+    - **Format**: ``schema://signal/<value>``
+    - **Example**: ``debug://signal/SIGKILL``
+    - **Description**: Inform the plugin to immediately send the signal
+      specified immediately after the command to itself. The available
+      signal values are denoted by `python signal module <https://docs.python.org/3/library/signal.html#signal.SIG_DFL>`_
+
+*SUCCESS* and *ERROR*
+    - **Format**: ``schema://command/(details)`` [Details are optional]
+    - **Examples**: ``debug://success``, ``debug://error/``, ``debug://error/general[foo=True;]``
+    - **Description**: Denotes whether a transferred file was successful or
+      failed due to an error. If any **error** commands are encoding in
+      the transfer URLs then the plugin will exit with ``1`` unless specified
+      otherwise or an internal failure occurred. Both of these commands
```

Comment:
```suggestion
      failed due to an error. If anu **error** command is om
      the transfer URL list then the plugin will exit with ``1`` unless specified
      otherwise or an internal failure occurred.  The last error command wins.  Both of these commands
```